hazelcast/include/hazelcast/client/serialization/pimpl/compact/compact_impl.h

```diff
@@ -604,10 +641,18 @@ template<typename T>
 void inline compact_stream_serializer::write(const T& object,
                                              object_data_output& out)
 {
-    const auto& schema_v = schema_of<T>::schema_v;
-    put_to_schema_service(schema_v);
-    out.write<int64_t>(schema_v.schema_id());
-    default_compact_writer default_writer(*this, out, schema_v);
+    const boost::optional<schema>& schema_v = class_to_schema<T>::get();
+
+    if (!schema_v.has_value()) {
```

Comment:
I think we need to carry the mutex in class_to_schema to this position. In current implementation there are two things exists. 1. schema_v is a reference and it is pushed to schemas_will_be_replicated_ vector, so it can be used at any time without locking mutex. 2. there is a performance decrease because in every write, the mutex will be locked and unlocked. I think, we first query has_value and if it does not contain anything, we should should lock the mutex and set the schema.

Reply:
Fixed.